packages/core/core/src/RequestTracker.js

```diff
@@ -0,0 +1,427 @@
+// @flow strict-local
+
+import type {AbortSignal} from 'abortcontroller-polyfill/dist/cjs-ponyfill';
+import type {File, FilePath, Glob} from '@parcel/types';
+import type {Event} from '@parcel/watcher';
+import type {NodeId} from './types';
+
+import invariant from 'assert';
+import nullthrows from 'nullthrows';
+import {isGlobMatch, md5FromObject} from '@parcel/utils';
+import Graph, {type GraphOpts} from './Graph';
+import {assertSignalNotAborted} from './utils';
+
+type SerializedRequestGraph = {|
+  ...GraphOpts<RequestGraphNode, RequestGraphEdgeType>,
+  invalidNodeIds: Set<NodeId>,
+  globNodeIds: Set<NodeId>,
+  unpredicatableNodeIds: Set<NodeId>
+|};
+
+type FileNode = {|id: string, +type: 'file', value: File|};
+type GlobNode = {|id: string, +type: 'glob', value: Glob|};
+export type Request = {|
+  id: string,
+  +type: string,
+  request: mixed,
+  result?: mixed
+|};
+
+type RequestNode = {|
+  id: string,
+  +type: 'request',
+  value: Request
+|};
+type RequestGraphNode = RequestNode | FileNode | GlobNode;
+
+type RequestGraphEdgeType =
+  | 'subrequest'
+  | 'invalidated_by_update'
+  | 'invalidated_by_delete'
+  | 'invalidated_by_create';
+
+const nodeFromFilePath = (filePath: string) => ({
+  id: filePath,
+  type: 'file',
+  value: {filePath}
+});
+
+const nodeFromGlob = (glob: Glob) => ({
+  id: glob,
+  type: 'glob',
+  value: glob
+});
+
+const nodeFromRequest = (request: Request) => ({
+  id: request.id,
+  type: 'request',
+  value: request
+});
+
+export class RequestGraph extends Graph<
```

Comment:
Should this be in its own file?

Reply:
Again, I was mostly keeping them together while I was trying things out. The `RequestTracker/RequestGraph/RequestRunner` are all pretty coupled though, so I could see value in keeping them together to understand quickly how they work together? I don't feel super strongly about that though.